clippy_lints/src/non_zero_suggestions.rs

```diff
@@ -0,0 +1,122 @@
+use clippy_utils::diagnostics::span_lint_and_sugg;
+use clippy_utils::source::snippet;
+use rustc_errors::Applicability;
+use rustc_hir::{Expr, ExprKind};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_middle::ty::{self, Ty};
+use rustc_session::declare_lint_pass;
+use rustc_span::symbol::sym;
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Checks for conversions from `NonZero` types to regular integer types,
+    /// and suggests using `NonZero` types for the target as well.
+    ///
+    /// ### Why is this bad?
+    /// Converting from `NonZero` types to regular integer types and then back to `NonZero`
+    /// types is less efficient and loses the type-safety guarantees provided by `NonZero` types.
+    /// Using `NonZero` types consistently can lead to more optimized code and prevent
+    /// certain classes of errors related to zero values.
+    ///
+    /// ### Example
+    /// ```no_run
+    /// use std::num::{NonZeroU32, NonZeroU64};
+    ///
+    /// fn example(x: u64, y: NonZeroU32) {
+    ///     // Bad: Converting NonZeroU32 to u64 unnecessarily
+    ///     let r1 = x / u64::from(y.get());
+    ///     let r2 = x % u64::from(y.get());
+    /// }
+    /// ```
+    /// Use instead:
+    /// ```no_run
+    /// use std::num::{NonZeroU32, NonZeroU64};
+    ///
+    /// fn example(x: u64, y: NonZeroU32) {
+    ///     // Good: Preserving the NonZero property
+    ///     let r1 = x / NonZeroU64::from(y);
+    ///     let r2 = x % NonZeroU64::from(y);
+    /// }
+    /// ```
+    #[clippy::version = "1.81.0"]
+    pub NON_ZERO_SUGGESTIONS,
+    restriction,
+    "suggests using `NonZero#` from `u#` or `i#` for more efficient and type-safe conversions"
+}
+
+declare_lint_pass!(NonZeroSuggestions => [NON_ZERO_SUGGESTIONS]);
+
+impl<'tcx> LateLintPass<'tcx> for NonZeroSuggestions {
+    fn check_expr(&mut self, cx: &LateContext<'tcx>, expr: &'tcx Expr<'tcx>) {
+        if let ExprKind::Call(func, [arg]) = expr.kind {
+            if let ExprKind::Path(qpath) = &func.kind {
```

Comment:
maybe check if the expr is a binary expression first? Therefore the below example doesn't cause error after applying lint suggestion:

```rust
let x: u64 = u64::from(NonZeroU32::new(5).unwrap());
```

Reply:
First I thought like that only, based on the suggestion from the issue but then I thought do we need it?

Reply:
Do we need to check for binary expr?

Reply:
I think yes? Notice that the tests are failing because the lint shouldn't suggestion a fix to a non-`binary expr`:

```console
error[E0308]: mismatched types
  --> tests/ui/non_zero_suggestions.fixed:68:5
   |
LL | fn no_bin_exp(x: u64, y: NonZeroU32) -> u64 {
   |                                         --- expected `u64` because of return type
LL |     NonZeroU64::from(y)
   |     ^^^^^^^^^^^^^^^^^^^ expected `u64`, found `NonZero<u64>`
   |
   = note: expected type `u64`
            found struct `std::num::NonZero<u64>`
help: call `Into::into` on this expression to convert `std::num::NonZero<u64>` into `u64`
   |
LL |     NonZeroU64::from(y).into()
   |                        +++++++
```

In this case, you shouldn't give a `MachineApplicable` suggestion at least, as it's telling `rustfix` to apply that suggestion because it's "always correct", but that's not the case since there could be type difference. I think the original issue specifically limit the cases to `binary expr` was because that they wanted to play safe, as the resulting type of an integer that `add|sub|div|mul|...` another `NonZero` type is still the integer type itself ([example](https://doc.rust-lang.org/std/primitive.u32.html#impl-Div%3CNonZero%3Cu32%3E%3E-for-u32)), so it can always being replaced to the `NonZero` type directly.

I think you could try adjusting the `applicability` base on whether the `expr` is binary or not, if yes, you can continue to offer a `MachineApplicable` suggestion as it is. But if not, you should state the applicability as `MaybeIncorrect`, so when a user runs clippy with `--fix`, it wouldn't cause any `suggestion-cause-error` issue. 
For example, you could try something like this:

```rust
let mut appllicability = Applicability::MaybeIncorrect;
if let ExprKind::Binary(_, _, rhs) = expr.kind {
    applicability = Applicability::MachineApplicable;
    check_for_fn(...);
} else {
    check_for_fn(...);
}
```

Reply:
I guess you are right. Limiting the lint to binary expressions would be better. It avoids complex cases and aligns with the original intent of the issue. I'll focus on implementing this change and add relevant test cases. I'll update the PR soon with these changes.